### PR TITLE
Generate RID graph in self-contained builds

### LINKED ISSUE #3361: Self-contained apps should include RID fallback graph in .deps.json
With the introduction of `AssemblyDependencyResolver` in 3.0 we're making it relatively easy to dynamically load components with their entire dependency trees. Such components can be portable in the sense that they include RID specific assets (for example different asset for Windows and Linux).

In order to correctly resolve dependencies of such component, the `AssemblyDependencyResolver` must have a RID fallback graph to follow (so that when we're running on `win10-x64`, we know that we should also look for assets which are `win-x64` or even `win`). The component itself should not carry the RID fallback graph with it (it would need to have the graph of all the RIDs there are and it would need to be updated with every addition).

Just like framework dependent apps which rely on the framework to provide the RID fallback graph, the component should rely on the host app to provide the RID fallback graph.

So if the host app is a framework dependent app, the RID fallback graph comes from the core framework `Microsoft.NETCore.App`. In this case everything works fine.

If the host app is a self-contained app, it in itself doesn't need a RID fallback graph since it's been built/published for one specific RID and thus it only carries assets which are necessary for that RID. But if a portable component is about to be loaded into such app, the RID fallback graph is necessary.

The proposal is to modify the SDK to include the RID fallback graph in self-contained app's `.deps.json`. It should be the graph filtered by the RID the app is targeting (just like what we do for the shared framework).

The problem is also discussed in https://github.com/dotnet/core-setup/issues/6960

### PR DESCRIPTION
**Pull request by swaroop-sridhar**
## Issue 
#3361

## Scenario
Customer tries to load load plugins with RID-specific assets into a self-contained app.
This operation currently fails.

## Fix

In order to support loading components (plugins) with RID-specific assets, the `AssemblyDependencyResolver` requires the RID fallback graph.

The component itself should not carry the RID fallback graph with it (it would need to have the graph of all the RIDs there are and it would need to be updated with every addition).

For framework dependent apps, the RID fallback graph comes from the core framework Microsoft.NETCore.App, so there is no need to write it into the app.

If self-contained apps, the (applicable subset of) RID fallback graph needs to be written to the deps.json manifest.

## Risk
Medium

This change affects contents generated in the deps.json file for every single-file app.
However, it only adds information (RID-graph) that is consumed by the `AssemblyDependencyResolver`.
It doesn't change any other sections in the `deps.json` manifest.

## Customer impact
Customer is able to load load plugins with RID-specific assets into a self-contained app.

### CODE REVIEW

**Comment by swaroop-sridhar** [comment]
@nguerrera  please review
CC: @vitek-karas

**Comment by swaroop-sridhar** [comment]
I've verified that the repro provided by @shvez in https://github.com/dotnet/coreclr/issues/25516 works fine after this fix.

File: src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenADependencyContextBuilder.cs
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenADependencyContextBuilder.cs]:
```suggestion
            DependencyContext dependencyContext = new DependencyContextBuilder(mainProject, projectContext, includeRuntimeFileVersions: false, runtimeGraph: null)
```

Extreme nit.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs]: What is "current" in this context. Does this work if cross-publishing to a different RID than build host?

Comment (swaroop-sridhar) [reply src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs]: Its the target RID for which the app is being built, I'll fix the comment wording. Thanks.

Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs]: Ah, I see the RID on project context. I was wondering where it would get it from.

**Review by nguerrera — APPROVED** [review]

**Comment by nguerrera** [comment]
Hold on merging for a bit until 3.1 Preview 2 is out.

**Review by nguerrera — CHANGES_REQUESTED** [review]
Waiting for branch to open

**Comment by nguerrera** [comment]
Looks like some miscommunication here on my part, this will need to go to tactics for approval.

**Comment by nguerrera** [comment]
Sorry for the trouble.

**Comment by wtgodbe** [comment]
@nguerrera was this approved for GA?

**Comment by nguerrera** [comment]
@wtgodbe I don't think so. @swaroop-sridhar Can you fill out the template and ask for approval from tactics? Otherwise we would have to hold this for 3.1.200.

**Comment by swaroop-sridhar** [comment]
@nguerrera  I have filled out the ask-mode template, but can't edit the labels on this PR.
Can you please add the ask-mode label? Thanks.

**Comment by swaroop-sridhar** [comment]
I've sent a note to tactics about this issue; it's probably OK if we can only take this change in 3.1.200. Thanks.

**Comment by swaroop-sridhar** [comment]
Moved this PR to 3.1.2xx branch: https://github.com/dotnet/sdk/pull/4067

**Pull request closed without merging**
